compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -176,6 +176,34 @@ object messages {
            |Or, add an explicit `()' as a parameter list to ${cdef.name}."""
   }
 
+  case class AnonymousFunctionMissingParamType(param: untpd.ValDef,
+                                               args: List[untpd.Tree],
+                                               tree: untpd.Function,
+                                               pt: Type)
+                                              (implicit ctx: Context)
+  extends Message(AnonymousFunctionMissingParamTypeID) {
+    val kind = "Syntax"
+
+    val msg = {
+      val ofFun =
+        if (MethodType.syntheticParamNames(args.length + 1) contains param.name)
+          i" of expanded function $tree"
+        else
+          ""
+
+      i"missing parameter type for parameter ${param.name}$ofFun, expected = $pt"
+    }
+
+    val explanation =
+      hl"""|Anonymous functions must define a type. For example, if you define a function like this one:
+           |
+           |val f = { case l@List(1,2,3) => Some(l) }
```

Comment:
For syntax highlighting this needs to be wrapped:

```diff
+           |${"val f = { case xs @ List(1, 2, 3) => Some(xs) }"}
-           |val f = { case l@List(1,2,3) => Some(l) }
```
Some spaces would also be awesome! 👍